### ISSUE TITLE
FEATURE: TV annotation win/lose/make a difference, based on polls

### ISSUE THREAD

**Issue by baob** [swapmyvote/swapmyvote#816]
Annotate a single TV recommendation (so, we know the constituency and the party that's going to get the vote in that constituency)

Feature is variously described. By @baob 

> inferences from polling data (losing/making a difference/winning) ?

By @aspiers 

> making the bar graph easier to understand, e.g. hiding it behind some text using maybe three categories of descriptions:
> * your preferred party is polling a lead of X% in this constituency,
> * polls say it's marginal / close call,
> * your preferred party is 2nd in the polls in this constituency, trailing by Y% (although it feels like we also need:
> * your preferred party is 3rd / 4th / ...

For first cut, I'm going to aim for basing this on the marginal score and (new) signed marginal score
* "(green tick) polling suggests your swap could make a difference for this party"
  - because marginal score < 10%, they're likely to just win or just lose
* "polling suggests your swap is supporting a safe win for this party"
   - because signed marginal score > 10%, positive they are way ahead
* "polling suggests your party will still lose despite your swap"
   - because signed marginal score < 10% negative, they are way behind

"this party" will probably be the actual party name. 

/cc @tdg

**Comment by aspiers** [comment]
@baob Great, thanks!

One of the things which came up in my UX discussions last Monday was that using a marginality score we construct ourselves without any explanation of how it works runs into the risk of SMV being perceived as "inventing" numbers which we could use to favour a particular outcome.  Of course this isn't true, but we need to make sure that we present this as clearly being based on hard numbers which we have no influence on.  So I think the text used here must refer directly to the difference in poll predictions between the poll leader and party in question, rather than to a score.

Another slight risk is editorialising the interpretation of the raw numbers.  For example, some people may consider "way ahead" to mean > 15% and others > 5%.  Even the choice of 10% (or whatever) may be considered partisan by people looking to attack us, although it's also quite possible I'm being too paranoid here -- but it's worth thinking about at least.

/cc @tdg as we really need you to chime in here.

**Comment by baob** [comment]
@tdg interested in your views on this. If we can't explain this sufficiently well to the voter, maybe we shouldn't do it ? 

The best explanation (IMHO) of 'marginal score', which this is based on, is in code comments here:  https://github.com/swapmyvote/swapmyvote/blob/master/app/models/poll.rb#L11

**Comment by aspiers** [comment]
Just to be clear, I think we should definitely do this and sorry if I did any accidental scaremongering!  Just pointing out it needs a bit of careful thought on the wording.

**Comment by baob** [comment]
@aspiers @tdg How about this wording.

"(green tick) check the poll chart, looks like your swap could make a difference for Labour"
"check the poll chart, looks like your swap may be supporting a safe win for this party"
"check the poll chart, looks like Labour may still lose despite your swap"

**Comment by aspiers** [comment]
I think it would be helpful to include hard figures, e.g.

- "(green tick) your swap could make a difference for Labour which is only trailing OTHER_PARTY by X% in the polls"
- "your swap may be supporting a safe win for Labour which is already leading by X% in the pools"
- "Labour is trailing by X% in the polls and may still lose despite your swap"

**Comment by aspiers** [comment]
Don't suppose there is a user-friendly frontend to STT's average polls data somewhere out there, which we could hyperlink to from these?  Otherwise we can just link to a new entry in our FAQ.

**Comment by tdg** [comment]
We might need a page about polls so we can link to it internally, and update with fresh info regularly

**Comment by tdg** [comment]
Some thoughts and copy WIP

TIGHT SECOND 
- (green tick) your swap could help (PREFERRED_PARTY). It is in second place (Y_PLACE if Y=2) by X% in the latest polls, behind OTHER_PARTY. Your swap would help them tighten or even win the seat. 

TIGHT FIRST 
- (green tick) your swap could help (PREFERRED_PARTY). It is leading place (Y_PLACE if Y=1) by X% in the latest polls, ahead of OTHER_PARTY. Your swap would help them tighten or even win the seat. 


TIGHT FIRST OR TIGHT SECOND WITH A LOT OF HEARTS
- as both above but with addendum, most tactical voting websites suggest a 'vote with your heart'

DISTANT SECOND
- (PREFERRED_PARTY) is in (Y_PLACE IF Y=2) by X% in the polls behind OTHER_PARTY and may still lose the seat despite your swap. The vote can still contribute to (party's) national share and even the local campaign for future elections. 

DISTANT THIRD OR WORSE
- (PREFERRED_PARTY) is in (Y_PLACE IF Y>=2) by X% in the polls behind OTHER_PARTY and OTHER_PARTY2 and is likely still to lose the seat despite your swap. The vote can still contribute to (party's) national share. 

SAFE PREFERRED
- your swap may help consolidate a safe seat for (PREFERRED_PARTY) which is in first place (Y_PLACE if Y=1) by X% in the polls, ahead of OTHER_PARTY. This can be a worthwhile swap overall if you are offering a swap in a hard fought seat for your partner's preferred party.

**Comment by aspiers** [comment]
@tdg This is great, thanks a lot! Just a quick thought: will we run into UX problems if this copy is on the longer side? I'm worried that it might be information overload for some or even many users. Didn't some seasoned campaigner recently give us advice that anything over 5 words long won't get read by most people? Can't remember where I read that, maybe on one of the WhatsApp groups or something?

**Comment by aspiers** [comment]
BTW how is this issue different from #667 ?

**Comment by tdg** [comment]
Yes I think we should aim to essentialise the copy over time. 

Some elements could be visual (leading or challenger party, say), and I could imagine a visual signifier for some value of marginality', but I think we need to work out the elements first before we boil it down.

**Comment by aspiers** [comment]
For me a large part of the struggle with the bar chart is that the parties aren't labelled (only visible via mouse hover), and I haven't perfectly internalised the various party colours.  It should be pretty easy to fix this, I guess.

**Comment by baob** [comment]
> TIGHT FIRST OR TIGHT SECOND WITH A LOT OF HEARTS

I can't see how this would ever happen. 'Heart' is an indicator that the TV-recommended party is way out in front with no chance of losing.

That's right isn't it ?

**Comment by baob** [comment]
> TIGHT SECOND
> TIGHT FIRST

Are we that confident about the poll predictions that it's worth adding this distinction ?

**Comment by tdg** [comment]
AIUI Heart can also be eg a progressive on progressive seat or safe seat (held by any party) where it doesn't matter how you vote so you can vote with your heart. True that the progressive TV sites would tend to make a recommendation in a Tory safe seat eg, but perhaps not if the progressive parties are evenly matched

**Comment by baob** [comment]
This is becoming very complicated. We don't record "heart" as a thing you can query on, we translate it into the individual parties.

**Comment by baob** [comment]
Also I don't think think you can necessarily infer heart means the same thing to each TV site. 

I'm wondering also if we're in danger of assuming lots of hearts mean "good for the voter" that depends entirely on the voters alignment with the campaigns.

**Comment by aspiers** [comment]
I think we're mixing two different things here: 

1. what the polls suggest in terms of what difference a swap may (or may not) be able to make
2. TV recommendations

If we try to annotate based on combinations of both, then the number of scenarios we have to deal with explodes in complexity, and I'm not convinced it's realistic to take this approach, at least not this close to the GE.  It also feels like this would increase the risk of editorializing recommendations.

So my hunch is that we should only annotate based on the polls and marginality scores.  I believe the discussion originally came from (or at least was reinvigorated by) my discussion with Lorelai about UX, in which I got the feedback that the relevance of the bar chart is not at all clear at the moment.  So the main point of these annotations would be to make the bar chart more understandable (or even replace it, although I'm personally not in favour of ditching it).

**Comment by aspiers** [comment]
It's getting quite hard to follow all the different strands of this discussion, as it seems to be spanning this and #667.  Are we all agreed that [they are duplicates of each other](https://github.com/swapmyvote/swapmyvote/issues/667#issuecomment-2179101137), albeit [with useful comments on both](https://github.com/swapmyvote/swapmyvote/issues/667#issuecomment-2179192944)?  If so, would be great to close one.  Or if not, to clarify what the differences are.

**Comment by baob** [comment]
Can I suggest close the other, older one ? And if there are any particularly useful comments there then copy them here ?

**Comment by baob** [comment]
> So my hunch is that we should only annotate based on the polls and marginality scores

I agree.

**Comment by baob** [comment]
> I think it would be helpful to include hard figures, e.g.
> 
> * "(green tick) your swap could make a difference for Labour which is only trailing OTHER_PARTY by X% in the polls"
> * "your swap may be supporting a safe win for Labour which is already leading by X% in the pools"
> * "Labour is trailing by X% in the polls and may still lose despite your swap"

Here's how this would look:

![image](https://github.com/swapmyvote/swapmyvote/assets/55932/6bf96ed9-9d24-4333-82c2-a8bb56f94c99)

**Comment by aspiers** [comment]
This is looking very promising, thanks!  I think we need some UI tweaks:

- The text at the bottom is divorced from the chart and easy to miss.  It's also inside the recommendations box even though it's nothing to do with the TV recommendations.  So I think it should go above or below the chart, probably just below "Does this help your vote count?".
- In that case we don't need the "Check the poll chart" text which also makes things more concise.
- Change " percent" to `%`.
- Change "despite your swap" to "despite this swap".

**Comment by baob** [comment]
@aspiers @tdg  How about this ?

![image](https://github.com/swapmyvote/swapmyvote/assets/55932/09812365-9952-4a95-8594-e5e1ed243349)

**Comment by aspiers** [comment]
Looks great!  I'm not convinced we need both "Does this help your vote count?" and "Polls interpretation", as they're both the same thing.  I suggest sticking with just the former, although maybe it's better placed inside the new box than outside it.

Only other thoughts currently is the choice of icon.  If the swap helps then I think a green tick is more suitable than a red exclamation mark, since the latter looks more like an error.  If the swap is unlikely to make a difference then we could use a question mark, but @tdg may want to weigh in here.

**Comment by baob** [comment]
"Does this help your vote count ?" was introduced for recommendations, not for polling interpretation, are you sure you want  to remove it as a header for recommendations ? 

I chose the red exclamation mark for polls as something distinctly different from the green tick for recommendations ... otherwise it just blurs in with recommendations. Happy to admit I'm not the UX guy though ... maybe we could just use different, brightly coloured blobs or something ? @stevebaxter may have ideas.

I'll wait till the other opinions are in before changing anything.

Possible icons  to use: 
<p>&#x26A1; &#x2B50; &#x1F15F; &#x1F161; &#x2728; &#x1F4CC;</p>

**Comment by aspiers** [comment]
Yes I think it's fine to change the header position. ⚖️ and 💪🏻 are also options.

**Comment by baob** [comment]
@tdg Can we have your take on this please ? "Does this help your vote count?" was text originally introduced by you to apply to TV recommendations. @aspiers is suggesting it should only apply to the poll interpretations (new feature being discussed here) and NOT to TV recommendations. 

Happy to make the change, so long as I know we have a consensus on what's needed :)  

> I'm not convinced we need both "Does this help your vote count?" and "Polls interpretation", as they're both the same thing. I suggest sticking with just the former, although maybe it's better placed inside the new box than outside it.

---------

@aspiers re

> I'm not convinced we need both

 I agree it could do with slimming down somewhat. Perhaps 'Polls interpretation' inside the box is a bit redundant, given the full sentence that follows.

**Comment by tdg** [comment]
> Perhaps 'Polls interpretation' inside the box is a bit redundant, given the full sentence that follows.

Agree

**Comment by tdg** [comment]
In order to reflect the changing nature of polls, please can we change 'already' to currently, and reduce the significant figures of the percentage, to make an integer or a single decimal place at most?

**Comment by tdg** [comment]
Please could we also add 'Tactical voting' to 'recommendations': 'Tactical voting recommendations'

**Comment by tdg** [comment]
Looking amazing folks!

**Comment by baob** [comment]
> reduce the significant figures of the percentage, to make an integer or a single decimal place at most?

Trying to think how to deal with percentages less than 0.5. Rounding down to zero and rounding up to 1 will both give problems.

Certainly need to add 'about' just before the figure anyway if we're rounding.

**Comment by tdg** [comment]
Fair point, single decimal would be good, could we round down if <0.05, up if 0.05>=

**Comment by baob** [comment]
Latest changes:

![image](https://github.com/swapmyvote/swapmyvote/assets/55932/a7fc96a1-8ff4-463b-b5e7-f7148c5daba2)

**Comment by aspiers** [comment]
[**@tdg**](https://github.com/tdg) commented on [June 25, 2024 4:48 PM](https://github.com/swapmyvote/swapmyvote/issues/816#issuecomment-2189315307 "Tue Jun 25 2024 16:48:19 GMT+0100 - Replied by Github Reply Comments"):
> Fair point, single decimal would be good, could we round down if <0.05, up if 0.05>=
> 
> _Originally posted by @tdg in https://github.com/swapmyvote/swapmyvote/issues/816#issuecomment-2189315307_

I think we need to use significant digits rather than fixed decimals:

| Actual % | Displayed |
|--------|--------|
| 2.65% | 3% |
| 0.49% | 0.5% |
| 0.0317% | 0.03% |
| 0.0058% | 0.006% |

**Comment by aspiers** [comment]
@baob I believe this can be accomplished via `"%.#1g" % percentage`:

```
$ irb
3.3.2 :001 > [2.65, 0.49, 0.317, 0.0058].map { |f| "%.1g%%" % f }
 => ["3%", "0.5%", "0.3%", "0.006%"] 
```

**Comment by baob** [comment]
Unfortunately completely broken for >= 10 ... I will apply selectively

```
% irb
3.3.2 :001 > [13.9, 18.2, 7.34, 2.65, 0.49, 0.317, 0.0058].map { |f| "%.1g%%" % f }
 => ["1e+01%", "2e+01%", "7%", "3%", "0.5%", "0.3%", "0.006%"]
```
